documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,296 @@
+
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    public class VersionInfo
+    {
+        public static const int MIN_SUPPORTED_VERSION;
+        public static const int MAX_SUPPORTED_VERSION;
+    }
+}
+```
+
+### DiagnosticsCommand
+This is an enum for all the diagnostics command supported in the current version of the library.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    /// <summary>
+    /// An enum for all the supported diagnostics IPC command
+    /// </summary>
+    public enum DiagnosticsCommand
+    {
+        Dump,
+        EventPipe
+        Profiler
+    }
+}
+```
+
+### DiagnosticsIpcHelper
+This is a class that contains some utility methods for various purposes.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    /// <summary>
+    /// A utility class that contain various helper methods to interact with the diagnostics IPC.
+    /// </summary>
+    public class DiagnosticsIpcHelper
+    {
+        /// <summary>
+        /// Get all the active processes that can be attached to.
+        /// </summary>
+        /// <returns>
+        /// IEnumerable of all the active process IDs.
+        /// </returns>
+        public static IEnumerable<int> GetActiveProcesses();
+
+        // The default path that the diagnostics client library looks for the IPC socket (default: /tmp)
+        public static string DefaultIpcSocketPath { get; set; } 
+
+        /// <summary>
+        /// Get all the supported commands by this version of the diagnostics IPC.
+        /// </summary>
+        /// <returns>
+        /// IEnumerable of all supported commands Command Code.
+        /// </returns>
+        public static IEnumerable<int> GetAvailableCommands();
+
+        /// <summary>
+        /// Convert a command name in the right format (CommandSet:Command) - i.e. EventPipe.CollectTracing - to corresponding command code.
+        /// </summary>
+        /// <returns>
+        /// Corresponding command code for the given command name. This may throw UnknownCommandException.
+        /// </returns>
+        public static int GetCommandCodeFromCommandName(string commandName);
+
+        /// <summary>
+        /// Convert a command code to its string representation
+        /// </summary>
+        /// <returns>
+        /// Corresponding command name for the given command code. This may throw UnknownCommandException
+        /// </returns>
+        public static string GetCommandNameFromCommandCode(int commandCode);
+
+    }
+}
+```
+
+### EventPipeCommandHandler
+This is a CommandHandler class for sending EventPipe commands across the diagnostics IPC channel.
+```cs
+namespace Microsoft.Diagnostics.Client
+{
+    /// <summary>
+    /// A class for handling EventPipe commands.
+    /// </summary>
+    public class EventPipeCommandHandler
```

Comment:
Would this library be the place for a higher-level abstraction of something like "session" for EventPipe?  For some of my GCDump work, I put together a crude session wrapper (see: https://github.com/josalem/diagnostics/blob/2e8e4f3576a45e51292bfd9c71cb80b272258d36/src/Tools/dotnet-trace/CommandLine/Commands/DotNetHeapDumper.cs#L422-L461), but I could see extrapolating that to have a builder class, be disposable, and abstract away some of the implementation details of the IPC protocol.  The "session" abstraction would be implemented via this Controller class.

Reply:
We chatted offline during coffee but reposting as a comment to open it up for discussion w/ wider audience :p 

The way I imagined this library is a minimal set of APIs for library authors to interact with the runtime via the diagnostics IPC protocol. I think some kind of session management *could* be useful but it depends on where we draw the line for something that all library authors would rely on. As far as I can tell your EventPipeSession class in the code you wrote for GCDump work, you keep a reference to an EventPipeEventSource instance, making it act as a higher level abstraction that handles both EventPipe specific commands as well as some of the event handling logic with EventPipeEventSource. By that point, we're bringing in dependency on the entire TraceEvent library for just EventPipe that some users of this library may not even use (i.e. someone who might be using it as a thin shim for profiler attach, or some dump trigger based on counter data, etc.), and that is the part I don't like. 

That being said, we already do have a dependency on TraceEvent (even though the surface is very thin and could be removed with some trivial work), so I'm not closing on the possibility of adding more higher level abstractions/features to enable easier use of the API. But at the moment I am leaning towards keeping the APIs thin and minimal, and throw more features at it if needed.